python-package/lightgbm/engine.py

```diff
@@ -202,6 +201,7 @@ def train(params, train_set, num_boost_round=100,
             booster.best_iteration = earlyStopException.best_iteration + 1
             evaluation_result_list = earlyStopException.best_score
             break
+    booster._reverse_update_params()
```

Comment:
this seems cannot solve the problem. 
for example, when constructing Booster fails, this line will not be called. So the parameters cannot be reverted.